### ISSUE TITLE
[windows] tables improperly sized on close.

### ISSUE THREAD

**Issue by evoskuil** [libbitcoin/libbitcoin-database#107]
```
01:47:23.080196 INFO [node] Please wait while the node is stopping...
01:47:23.086175 DEBUG [network] Manual channel stopped: success
01:47:23.086175 DEBUG [network] Suspended manual connection.
01:47:23.087174 DEBUG [network] Stopped block_in protocol for [127.0.0.1:8333].
01:47:23.087174 DEBUG [network] Stopped block_out protocol for [127.0.0.1:8333].
01:47:23.087174 DEBUG [network] Stopped transaction_in protocol for [127.0.0.1:8333].
01:47:23.087174 DEBUG [network] Stopped transaction_out protocol for [127.0.0.1:8333].
01:47:23.090206 DEBUG [network] Valid block payload from [127.0.0.1:8333] (998068 bytes)
01:47:23.381390 INFO [network] Channel stopped: true
01:47:23.833820 DEBUG [database] Unmapped: "d:\blockchain-in\block_table" [4189020633]
01:47:23.836823 DEBUG [database] Unmapped: "d:\blockchain-in\block_index" [3294780]
01:48:00.980204 DEBUG [database] Unmapped: "d:\blockchain-in\transaction_table" [77878224980]
01:48:00.980705 INFO [node] Node stopped successfully.
```
```
01:55:36.472580 DEBUG [node] ================= startup 01/31/17 17:55:36 ==================
01:55:36.472580 INFO [node] ================= startup 01/31/17 17:55:36 ==================
01:55:36.479585 WARNING [node] ================= startup 01/31/17 17:55:36 ==================
01:55:36.488591 ERROR [node] ================= startup 01/31/17 17:55:36 ==================
01:55:36.497599 FATAL [node] ================= startup 01/31/17 17:55:36 ==================
01:55:36.502601 INFO [node] Using config file: "C:\ProgramData\libbitcoin\bn-blocks-in.cfg"
01:55:36.509607 INFO [node] Press CTRL-C to stop the node.
01:55:36.513609 INFO [node] Please wait while the node is starting...
01:55:36.521151 DEBUG [database] Buckets: block [650000], transaction [110000000], spend [250000000], history [107000000]
01:55:36.532669 DEBUG [database] Mapping: "d:\blockchain-in\block_table" [4189020632] (4096)
01:55:36.532669 DEBUG [database] Mapping: "d:\blockchain-in\block_index" [3294780] (4096)
01:55:36.532669 ERROR [node] Failure starting blockchain.
01:55:36.535666 ERROR [node] Node failed to start with error, operation failed.
01:55:36.539669 INFO [node] Please wait while the node is stopping...
01:55:36.765422 DEBUG [database] Unmapped: "d:\blockchain-in\block_table" [4189020632]
01:55:36.767894 DEBUG [database] Unmapped: "d:\blockchain-in\block_index" [3294780]
01:55:36.767894 INFO [node] Node stopped successfully.
```
**4189020632** vs. **4189020633**

It appears that the operating system truncates a single byte from the file.

This issue has been seen before but so far only on Windows. The underlying file map is platform-specific.

**Comment by evoskuil** [comment]
This has also been observed on `stealth_rows`:
```
09:30:08.175011 DEBUG [network] Manual channel stopped: success
09:30:08.175011 DEBUG [network] Suspended manual connection.
09:30:08.175011 DEBUG [network] Stopped block_in protocol for [139.59.210.142:8333].
09:30:08.175011 DEBUG [network] Stopped block_out protocol for [139.59.210.142:8333].
09:30:08.175011 DEBUG [network] Stopped transaction_in protocol for [139.59.210.142:8333].
09:30:08.175011 DEBUG [network] Stopped transaction_out protocol for [139.59.210.142:8333].
09:30:08.221915 DEBUG [database] Unmapped: "d:\server\blockchain\block_table" [6299560045]
09:30:08.221915 DEBUG [database] Unmapped: "d:\server\blockchain\block_index" [3617948]
09:30:11.925169 DEBUG [database] Unmapped: "d:\server\blockchain\transaction_table" [115081554420]
09:30:13.263837 DEBUG [database] Unmapped: "d:\server\blockchain\spend_table" [1717490472]
09:30:14.381619 DEBUG [database] Unmapped: "d:\server\blockchain\history_table" [603877976]
09:30:14.397267 DEBUG [database] Unmapped: "d:\server\blockchain\history_rows" [1689414019]
09:30:14.397267 DEBUG [database] Unmapped: "d:\server\blockchain\stealth_rows" [246196]
09:30:14.397267 INFO [node] Server stopped successfully.
```
```
09:30:48.784253 INFO [server] Please wait while the server is starting...
09:30:48.784253 DEBUG [database] Buckets: block [650000], transaction [110000000], spend [250000000], history [107000000]
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\block_table" [6299560045] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\block_index" [3617948] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\transaction_table" [115081554420] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\spend_table" [1717490472] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\history_table" [603877976] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\history_rows" [1689414019] (4096)
09:30:48.852814 DEBUG [database] Mapping: "d:\server\blockchain\stealth_rows" [246195] (4096)
09:30:48.852814 ERROR [node] Failure starting blockchain.
```

**246195** vs. **246196** (also 1 byte short).

**Comment by evoskuil** [comment]
Based on the assumption that this is a bug in the Windows memory mapped file implementation we can save an extra arbitrary byte, which will be ignored if not truncated.

**Comment by evoskuil** [comment]
This has not bee reproduced in over two months of continuous and dynamic operations on Windows. At this point we can reasonably assume it was a side effect of another issue that has been resolved.

**Comment by evoskuil** [comment]
Still here:
```
06:27:08.196804 DEBUG [database] Unmapped: "d:\server\blockchain\block_table" [2653848983]
06:27:08.370930 DEBUG [database] Unmapped: "d:\server\blockchain\block_index" [2973516]
06:28:56.858253 DEBUG [database] Unmapped: "d:\server\blockchain\transaction_table" [48257495656]
06:29:18.075754 DEBUG [database] Unmapped: "d:\server\blockchain\spend_table" [17000507932]
06:29:28.993141 DEBUG [database] Unmapped: "d:\server\blockchain\history_table" [3000068920]
06:29:34.625789 DEBUG [database] Unmapped: "d:\server\blockchain\history_rows" [38309327284]
06:29:34.801125 DEBUG [database] Unmapped: "d:\server\blockchain\stealth_rows" [854224]
06:29:34.801125 INFO [node] Server stopped successfully.
```
```
06:30:36.970219 DEBUG [database] Mapping: "d:\server\blockchain\block_table" [2653848983] (4096)
06:30:36.970219 DEBUG [database] Mapping: "d:\server\blockchain\block_index" [2973516] (4096)
06:30:36.970219 DEBUG [database] Mapping: "d:\server\blockchain\transaction_table" [48257495656] (4096)
06:30:36.971219 DEBUG [database] Mapping: "d:\server\blockchain\spend_table" [17000507932] (4096)
06:30:36.971219 DEBUG [database] Mapping: "d:\server\blockchain\history_table" [3000068919] (4096)
06:30:36.971219 DEBUG [database] Mapping: "d:\server\blockchain\history_rows" [38309327284] (4096)
06:30:36.971219 ERROR [node] Failure starting blockchain.
```
**3000068919** vs. **3000068920** (also 1 byte short).

**Comment by evoskuil** [comment]
The problem above is recorded on three different files, so clearly this is related to the underlying memory mapping: `memory_map.cpp`, `mman.c` or the `Windows API`.

**Comment by evoskuil** [comment]
After adding logging and additional error handling... Both show that the physical file size is larger than the logical size at the time of truncation, and that the truncation size specified is the correct logical size (this has been verified in the debugger).

```
23:35:41.011502 DEBUG [database] Unmapped: "d:\server\blockchain\block_table" [2612808684, 3285300237]
23:35:41.110902 DEBUG [database] Unmapped: "d:\server\blockchain\block_index" [2960004, 4247430]
23:36:55.356525 DEBUG [database] Unmapped: "d:\server\blockchain\transaction_table" [47496894744, 58548700387]
23:37:05.313473 DEBUG [database] Unmapped: "d:\server\blockchain\spend_table" [16732364276, 20802165714]
23:37:16.746676 DEBUG [database] Unmapped: "d:\server\blockchain\history_table" [2952145856, 3742726242]
23:37:40.990499 DEBUG [database] Unmapped: "d:\server\blockchain\history_rows" [37701078079, 45329657188]
23:37:40.995745 DEBUG [database] Unmapped: "d:\server\blockchain\stealth_rows" [792492, 879204]
23:37:40.995745 INFO [node] Server stopped successfully.
```
```
23:46:18.676643 DEBUG [database] Mapping: "d:\server\blockchain\block_table" [2612808684] (4096)
23:46:18.676643 DEBUG [database] Mapping: "d:\server\blockchain\block_index" [2960004] (4096)
23:46:18.677672 DEBUG [database] Mapping: "d:\server\blockchain\transaction_table" [47496894744] (4096)
23:46:18.677672 DEBUG [database] Mapping: "d:\server\blockchain\spend_table" [16732364276] (4096)
23:46:18.677672 DEBUG [database] Mapping: "d:\server\blockchain\history_table" [2952145855] (4096)
23:46:18.687683 DEBUG [database] Mapping: "d:\server\blockchain\history_rows" [37701078079] (4096)
23:46:18.687683 ERROR [node] Failure starting blockchain.
```

**Comment by evoskuil** [comment]
In the above example, the file system reports the file size after the store close as 2,952,145,856 bytes, which is the size reportedly written and which is expected at open (via the debugger).

![image](https://cloud.githubusercontent.com/assets/1369335/24984461/6d39c28c-1fa2-11e7-9b1f-957e3d369883.png)

After failing to open the store the implementation closes it, resulting in the one byte reduction of its size, matching the size reported in that log for both open and close.

![image](https://cloud.githubusercontent.com/assets/1369335/24984528/cb033556-1fa2-11e7-9521-b36ade046bdc.png)

In other words, the file is correctly sized on disk but the opening of the file fails to register the correct size, with the resulting behavior being expected.

**Comment by evoskuil** [comment]
By taking a snapshot of the store before attempting to reopen, we confirmed the above and were able to retry opening copies of the same store multiple times. The first attempt failed (above), and it was just coincidental that we had the snapshot. Subsequent attempts from the snapshot also failed. So even though the file is reported with the correct size, when the operating system opens the file the size it reports is less one byte.

**Comment by evoskuil** [comment]
The debugger confirms that `_fstat64(...)` returns a structure with the incorrect size for the file (one byte less than written and than what the Windows Explorer reports). Simply opening the file reduces its reported size by the one byte.

**Comment by evoskuil** [comment]
Turns out that [_wsopen()](https://msdn.microsoft.com/en-us/library/z0kc8e3z(v=vs.120).aspx) requires a special flag to prevent it from mutating the file. This problem is resolved by adding the `_O_BINARY` flag.
